File: 347. Top K Frequent Elements/Solution.cs
public class Solution {
    public int[] TopKFrequent(int[] nums, int k) {
        IDictionary<int, int> count = new Dictionary<int, int>();
        IList<int>[] frequencyBuckets = new List<int>[nums.Length + 1];
        for (int i = 0; i < frequencyBuckets.Length; i++) {
             frequencyBuckets[i] = new List<int>();
        }

        foreach (int num in nums) {
            if (!count.ContainsKey(num)) {
                count[num] = 0;
            }
            count[num]++;
        }

        foreach (KeyValuePair<int, int> entry in count) {
            frequencyBuckets[entry.Value].Add(entry.Key);
        }

        int[] topK = new int[k];
        int index = 0;
        for (int i = frequencyBuckets.Length - 1; i > 0; i--) {
            foreach (int num in frequencyBuckets[i]) {
                topK[index] = num;
                index++;
                if (index == k) {
                    return topK;
                }
            }
        }
        return topK;
    }
}

// Time complexity: O(n)
// Space complexity: O(n)

File: 347. Top K Frequent Elements/solution.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
    public:
        vector<int> topKFrequent(vector<int>& nums, int k) {
            unordered_map<int, int> count;
            vector<vector<int>> frequencyBuckets(nums.size() + 1);
    
            for (int num: nums) {
                count[num] = 1 + count[num];
            }
    
            for (const auto& entry: count) {
                frequencyBuckets[entry.second].push_back(entry.first);
            }
    
            vector<int> topK;
    
            for (int i = frequencyBuckets.size() - 1; i > 0; i--) {
                for (int num: frequencyBuckets.at(i)) {
                    topK.push_back(num);
                    if (topK.size() == k) {
                        return topK;
                    }
                }
            }
    
            return topK;
        }
};

// Time Complexity: O(n)
// Space Complexity: O(n)